File: akf_hrms/akf_hrms/report/late_arrival_report/late_arrival_report.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Late Arrival Report"] = {
	"filters": [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"reqd": 1
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"reqd": 1
		},
		{
			"fieldname":"employee",
			"label": __("Employee"),
			"fieldtype": "Link",
			"options": "Employee"
		},
		{
			"fieldname":"department",
			"label": __("Department"),
			"fieldtype": "Link",
			"options": "Department"
		}
	],

	// "formatter": function(value, row, column, data, default_formatter) {
	// 	value = default_formatter(value, row, column, data);

	// 	let class_name = "";
	// 	if (value == "P") class_name = "span-present";
	// 	else if (value == "A") class_name = "span-absent";
	// 	else if (value == "L") class_name = "span-leave";
	// 	else if (value == "Late") class_name = "span-late";
	// 	else if (value == "Early") class_name = "span-early";
	// 	else if (value == "Extra") class_name = "span-extra";
	// 	else if (parseFloat(data["Total hours worked"]) > parseFloat(data["Total hours assigned"]) && column.id == "Total hours worked") class_name = "span-extratime";
	// 	else if (parseFloat(data["Total hours worked"]) < parseFloat(data["Total hours assigned"]) && column.id == "Total hours worked") class_name = "span-lessime";
	// 	else if (data["Check In"] > data["Shift Check In"] && column.id == "Check In") class_name = "span-chk_in";

	// 	value = $(`<span class="${class_name}">${value}</span>`);
	// 	return value.wrap("<p></p>").parent().html();
	// },

	// "onload": function(frm) {
	// 	console.log(frm.report_name);
	// 	frm.page.add_inner_button(__("Send Email"), function() {
	// 		frappe.call({
	// 			method: "erpnext.hr.report.api_send_email.send_email",
	// 			args: {
	// 				from_date: frm.get_values().from_date,
	// 				to_date: frm.get_values().to_date,
	// 				employee: frm.get_values().employee,
	// 				company: frm.get_values().company,
	// 				department: frm.get_values().department,
	// 				// sub_department: frm.get_values().sub_department,
	// 				report_name: frm.report_name
	// 			}/*,
	// 			callback: function(r) {
	// 				frappe.set_route("Form", "Budget Invoice", r.message)
	// 			},*/
	// 		});
	// 	});
	// }
}

File: akf_hrms/akf_hrms/report/late_arrival_report/late_arrival_report.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters=None):
	columns, data = [], []

	columns = get_columns(filters)

	data = get_data(filters)

	return columns, data


def get_columns(filters):
	return [
		_("Employee ID") + ":Link/Employee:180", 
		_("Employee Name") + ":Data:200",
		_("Designation")+ ":Data:180",
		_("Department")+ ":Data:180",
		_("Location")+ ":Data:150",
		_("Gender")+ ":Data:150",
		_("Attendance Date")+ ":Data:150",
		_("Status")+ ":Data:150",
		_("Shift Check In")+ ":Data:180", 
		_("Shift Check Out")+ ":Data:180",
		_("Check In")+ ":Data:180", 
		_("Check Out")+ ":Data:180",
		_("Total hours assigned")+ ":Data:150", 
		_("Total hours worked")+ ":Data:150",
		_("Check In/Out Status")+ ":Data:150"
	]

def get_data(filters):
	return_list = []
	conditions, filters = get_conditions(filters)
	emp_info = []
	if ('HR Manager') in frappe.get_roles(frappe.session.user):
		emp_info = frappe.db.sql("""
			select att.employee, att.employee_name, att.custom_designation, att.department,
			att.custom_branch, att.custom_gender, att.attendance_date,
			(CASE WHEN att.status = 'Present' THEN 'P'
				WHEN att.status = 'Absent' THEN 'A'
				WHEN att.status = 'On Leave' THEN 'L' 
			END),
			IFNULL(SUBSTRING(att.custom_start_time, 12, 8),""), IFNULL(SUBSTRING(att.custom_end_time, 12, 8),""), 
			att.custom_total_working_hours, att.custom_hours_worked,
			(CASE
			WHEN att.late_entry = 1 AND att.early_exit = 1 THEN 'Late Entry and Early Exit'
			WHEN att.late_entry = 1 THEN 'Late Entry'
			WHEN att.early_exit = 1 THEN 'Early Exit'
			ELSE 'On Time'
			END)
			from `tabAttendance` as att
			where att.docstatus=1 %s
			""" % conditions, filters)

	else:
		get_emp_from_user_permission = frappe.db.sql("""select Distinct(for_value) from `tabUser Permission` where user=%s and 
								allow='Employee' """,(frappe.session.user))
		if get_emp_from_user_permission:
			for emp_id in get_emp_from_user_permission:
				emp_info = frappe.db.sql("""
				select att.employee, att.employee_name, att.custom_designation, att.department,
				att.custom_branch, att.custom_gender, att.attendance_date,
				(CASE WHEN att.status = 'Present' THEN 'P'
						WHEN att.status = 'Absent' THEN 'A'
						WHEN att.status = 'On Leave' THEN 'L' 
					END),
				IFNULL(SUBSTRING(att.custom_start_time, 12, 8),""), IFNULL(SUBSTRING(att.custom_end_time, 12, 8),""), 
				att.custom_total_working_hours, att.custom_hours_worked,
				(CASE
					WHEN att.late_entry = 1 AND att.early_exit = 1 THEN 'Late Entry and Early Exit'
					WHEN att.late_entry = 1 THEN 'Late Entry'
					WHEN att.early_exit = 1 THEN 'Early Exit'
					ELSE 'On Time'
				END)
				from `tabAttendance` as att
				where att.late_entry = 1 and att.docstatus=1 and att.employee=%s and att.attendance_date between %s and %s
				""", (emp_id[0], filters.get("from_date"), filters.get("to_date")))

				return_list.extend(emp_info)

	return return_list

def get_conditions(filters):
	conditions = ""
	if filters.get("from_date") and filters.get("to_date"): conditions += " and att.attendance_date between %(from_date)s and %(to_date)s"
	if filters.get("company"): conditions += " and att.company = %(company)s"
	if filters.get("employee"): conditions += " and att.employee = %(employee)s"
	if filters.get("department"): conditions += " and att.department = %(department)s"

	return conditions, filters
